test/cql-pytest/test_alter_table.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2024-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+# Tests for alter table statement
+
+import time
+import pytest
+
+from util import new_test_table, unique_name
+from nodetool import flush
+
+# Test checks only case of preparing `ALTER TABLE ... DROP ... USING TIMESTAMP ?` statement.
+# It is `scylla_only` because Cassandra doesn't allow to prepare the statement with ? as the timestamp.
+# More tests about `ALTER TABLE ... DROP ...` are in cassandra_tests/validation/operation/alter_test.py
+def testDropColumnWithTimestampPrepared(scylla_only, cql, test_keyspace):
+    with new_test_table(cql, test_keyspace, "id int, c1 int, v1 int, todrop int, PRIMARY KEY (id, c1)") as table:
+        prepared = cql.prepare(f"ALTER TABLE {table} DROP todrop USING TIMESTAMP ?")
+        for i in range(5):
+            cql.execute(f"INSERT INTO {table} (id, c1, v1, todrop) VALUES (1, {i}, {i}, {i}) USING TIMESTAMP {10000 * i}")
+
+        # Because the tests are run in parallel, we cannot be sure the table will stay in memory.
+        # It's safer to flush the table before dropping the column
```

Comment:
I think that this comment omits some of the context from our previous discussions. Why, in the first place, are we concerned here whether data is in memory or is flushed to disk?

Reply:
The comment is not really correct - we do **not** currently run tests in parallel on the same Scylla. But even if we don't run other tests in parallel, we still can't assume that that the table will stay in memory but....

I think this whole comment is wrong. You are claiming that you need to flush because Scylla may flush automatically, which.... doesn't make sense? If flushing is a problem how does flushing solve it? I think this test had a flush because of two reasons that act together:
1. The author was worried that Cassandra / Scylla might **not** flush the table if we don't do it explicitly.
2. The author was worried that if we don't flush, the result will be wrong.

Reason 1 is correct and reasonable - there is a high chance that the table will be in memtable only if not flushed. But I still don't understand how, if reason 2 is valid, how did the test not fail without the flush?

To cut a long story short, please remove the incorrect comment and then we can merge this test because @mykaul wants it in quickly.

Then please note for future reference that I'm troubled by the fact that I don't understand how despite the claim that this test should have failed without a flush, it actually passes. It may mean we don't understand something in this feature or this test. If @mykaul wasn't in a hurry, I would have insisted that we understand this before merging.

Reply:
About tests running in parallel - apparently we got that wrong, but the fact that the tests generate unique names for all schema items gave us this impression.

Let me provide the missing context right away, then - I agree that it should have been put into the comment:

As for flushing, the test is sensitive to _the act of flushing itself_.
I explained why the original Cassandra tests that we based the new tests on used to do it: https://github.com/scylladb/scylladb/pull/18115#discussion_r1570051825
In parallel to me posting the comment, @Jadw1 removed the flush and it turned out that this test keeps working. As to why - not 100% sure, but our theory is that if you drop a column then this column's data won't get automatically evicted, it only becomes inaccessible. However, if you use a particular timestamp to drop the column and then re-add it, apparently this data will become available again. Contents of a column can be dropped during flush or compaction, so that's the concern.

As for the "safer" argument: AFAIK we don't have 100% control about whether a flush occurs or not; to my knowledge if the size of memtables reach some particular threshold then they will be flushed, similarly commitlog can cause a flush if it grows too large. We thought that with parallel tests (that we thought are happening), this test case would be prone to flakiness because an uncontrolled flush might happen (as other tests share memory and maybe fo flushes themselves). If we flush ourselves and then do the desired schema changes, we avoid this problem completely.

Perhaps to be ultra safe we could disable compactions as well, though I suppose the risk here is even lower.

Reply:
The test is sensitive to time when potential flush might occur.

Logically test steps are:
1. insert some data
2. drop column
3. re-add column using timestamp, so part of the data in re-added column is discarded and the other part is still present in the table

If the flush never happens, that's totally okay for us and the test should pass.
However, if the table was flushed between steps 2 and 3, all data from dropped column would be discarded and not saved to sstable, cause it's not present in the table. Then after re-adding the column, all rows would have null in that field.

> AFAIK we don't have 100% control about whether a flush occurs or not; to my knowledge if the size of memtables reach some particular threshold then they will be flushed
From my knowledge, I also think it's true. So for safety, we can flush the data before dropping the column, so all data from the column is saved in sstable and we won't lose it.

Reply:
> About tests running in parallel - apparently we got that wrong, but the fact that the tests generate unique names for all schema items gave us this impression.

Just some historical context:

1. One of the goals of the cql-pytest framework is to be able to run on an instance of Scylla it can't control - perhaps some docker image the user ran, or perhaps even a remote Scylla Cloud cluster or even Amazon Keyspace or whatever.  Given that, tests should use a unique keyspace name (and not, say, "ks"), because they have no control over which other keyspaces exist.
2. Even if the tests do not run in parallel, we can have hundreds of tests written separately and we don't want a silly bug in one test (like forgetting to delete the table at the end of a test) to break another random test. So tests that create a table should use a unique name (`unique_name()`) and not something like "tbl".
3. By the way, ideally, many tests should share the same table (as a fixture) to reduce the test time. In that case, each test should use a unique partition in the shared table to avoid tests affecting each other. We have `unique_key_int()` et al. to help with that.
4. Finally, I always wanted to be able to run multiple tests in parallel against the same instance of Scylla (https://github.com/scylladb/scylladb/issues/14387) but it's not trivial, and nowadays we have a different solution for the same problem: test.py simply runs several separate instances of Scylla, and runs pytest on separate files against different Scyllas, in parallel. But this means that there is no parallelism in the single pytest source file running against one Scylla instance. But this is good enough for now and we have no plans to change it.